pkg/idtools/idtools_unix.go

```diff
@@ -265,13 +265,22 @@ func NewIdentityMapping(name string) (*IdentityMapping, error) {
 		return nil, err
 	}
 
-	subuidRanges := append(subuidRangesWithUserName, subuidRangesWithUID...)
-	subgidRanges := append(subgidRangesWithUserName, subgidRangesWithUID...)
+	var subuidRanges ranges
+	var subgidRanges ranges
 
-	if len(subuidRanges) == 0 {
+	if len(subuidRangesWithUID) > 0 {
+		subuidRanges = subuidRangesWithUID
+	} else if len(subuidRangesWithUserName) > 0 {
+		subuidRanges = subuidRangesWithUserName
+	} else {
```

Comment:
Actually wondering why we're looking up both; perhaps instead we should return early (try lookup sub-uids by UID -> if error, or no results -> lookup by Name -> repeat for sub-gids).

That said; ISTR there was some discussion about wether "user-name" or "user-id" takes precedence (user-name should probably not be able to take precedence (i.e., a user _named_ `0` should not be able to "trump" `root`), but linux disagreed on that (some discussion on https://github.com/moby/moby/issues/21436 and https://github.com/opencontainers/runc/commit/0a5293fa4ee5b38d1eeaa041754fd01f15960ad9 (https://github.com/opencontainers/runc/pull/708))

Reply:
Did a quick try at what it _could_ look like (feedback welcome) earlier;

```go
// NewIdentityMapping takes a requested username and
// using the data from /etc/sub{uid,gid} ranges, creates the
// proper uid and gid remapping ranges for that user/group pair
func NewIdentityMapping(name string) (*IdentityMapping, error) {
	usr, err := LookupUser(name)
	if err != nil {
		return nil, fmt.Errorf("Could not get user for username %s: %v", name, err)
	}

	subuidRanges, err := lookupSubUIDRanges(usr)
	if err != nil {
		return nil, err
	}
	subgidRanges, err := lookupSubGIDRanges(usr)
	if err != nil {
		return nil, err
	}

	return &IdentityMapping{
		uids: subuidRanges,
		gids: subgidRanges,
	}, nil
}

func lookupSubUIDRanges(usr user.User) ([]IDMap, error) {
	r, err := parseSubuid(strconv.Itoa(usr.Uid))
	if err != nil {
		return nil, err
	}
	if len(r) == 0 {
		r, err = parseSubuid(usr.Name)
		if err != nil {
			return nil, err
		}
	}
	if len(r) == 0 {
		return nil, errors.Errorf("no subuid ranges found for user %q", usr.Name)
	}
	return createIDMap(r), nil
}

func lookupSubGIDRanges(usr user.User) ([]IDMap, error) {
	r, err := parseSubgid(strconv.Itoa(usr.Uid))
	if err != nil {
		return nil, err
	}
	if len(r) == 0 {
		r, err = parseSubgid(usr.Name)
		if err != nil {
			return nil, err
		}
	}
	if len(r) == 0 {
		return nil, errors.Errorf("no subgid ranges found for user %q", usr.Name)
	}
	return createIDMap(r), nil
}
```

Reply:
I think it makes sense for UIDs to take precedence, as 99% of the time a numerical value is going to be a UID rather than a username.

I was thinking originally to implement separate functions, I wasn't sure if it was too much boilerplate but semantically it looks better to me.

Do you want me to update the commit with your implementation?

Reply:
Yes, there's definitely some boilerplating; I initially kept it inline, and, well, it was still boilerplating, so I extracted it to separate functions.

Definitely feel free to copy my code (check if I didn't make a mistake, as it was a quick write up 😅)